### PR TITLE
core[minor]: Support all versions of pydantic base model in argsschema

### PR DESCRIPTION
**Pull request by eyurtsev**
This adds support to any pydantic base model for tools.

The only potential issue is that `get_input_schema()` will not always return a v1 base model.

### CODE REVIEW

File: libs/core/langchain_core/tools.py [hunk 1]
```diff
@@ -119,6 +125,12 @@ def _create_subset_model(
     fn_description: Optional[str] = None,
 ) -> Type[BaseModel]:
     """Create a pydantic model with only a subset of model's fields."""
+    if PYDANTIC_MAJOR_VERSION == 2:
+        from pydantic.v1 import create_model  # pydantic: ignore
+    else:
+        from pydantic import (  # type: ignore[no-redef] # pydantic: ignore
+            create_model,
+        )
```
Comment (baskaryan) [inline libs/core/langchain_core/tools.py:133]: why not `from langchain_core.pydantic_v1 import create_tool`?

Comment (baskaryan) [reply libs/core/langchain_core/tools.py:133]: ie use the create_model_v1 that's already imported

File: libs/core/langchain_core/tools.py [hunk 2]
```diff
@@ -138,6 +150,75 @@ def _create_subset_model(
     return rtn
 
 
+def _create_subset_model_2(
```
Comment (baskaryan) [inline libs/core/langchain_core/tools.py:153]: nit: v2

File: libs/core/langchain_core/tools.py [hunk 3]
```diff
@@ -138,6 +150,75 @@ def _create_subset_model(
...
+        from pydantic.v1 import BaseModel as BaseModelV1  # pydantic: ignore
+
+        if issubclass(model, BaseModelV1):
+            return _create_subset_model_v1(
+                name,
+                model,
+                field_names,
+                descriptions=descriptions,
+                fn_description=fn_description,
+            )
```
Comment (baskaryan) [inline libs/core/langchain_core/tools.py]:
```suggestion
from langchain_core.pydantic_v1 import BaseModelV1Like

    if isinstance(model, BaseModelV1Like):
        return _create_subset_model_v1(
            name,
            model,
            field_names,
            descriptions=descriptions,
            fn_description=fn_description,
        )
    else:
      ...
```

File: libs/core/langchain_core/tools.py [hunk 4]
```diff
@@ -403,6 +380,16 @@ class ChildTool(BaseTool):
     two-tuple corresponding to the (content, artifact) of a ToolMessage.
     """
 
+    def __init__(self, **kwargs: Any) -> None:
+        """Initialize the tool."""
+        if "args_schema" in kwargs and kwargs["args_schema"] is not None:
+            if not is_basemodel_subclass(kwargs["args_schema"]):
+                raise SchemaAnnotationError(
```
Comment (baskaryan) [inline libs/core/langchain_core/tools.py:387]: is this a SchemaAnnotationError, sounds like a type error or value error?

Comment (baskaryan) [reply libs/core/langchain_core/tools.py:387]: do we need to update line 311 as well?

Comment (eyurtsev) [reply libs/core/langchain_core/tools.py]: Probably not in this PR for line 311 I don't know what its role was. Users could be relying on the identity of the exception

File: libs/core/langchain_core/utils/pydantic.py [hunk 5]
```diff
@@ -64,3 +128,106 @@ def wrapper(cls: Type[BaseModel], values: Dict[str, Any]) -> Dict[str, Any]:
         return func(cls, values)
 
     return wrapper
+
+
+def _create_subset_model_v1(
```
Comment (tdiggelm) [inline libs/core/langchain_core/utils/pydantic.py:133]: Unfortunately this function fails for pydantic v1 models that use `Annotated` with `Field`, e.g.

```python
class InputModel(BaseModel):
    query: Annotated[str, pydantic_v1.Field(description="Hello World")]

_create_subset_model_v1("test", InputModel, InputModel.__annotations__.keys())
```

This produces the following error:

```plain
ValueError: cannot specify `Annotated` and value `Field`s together for 'query'
```

**Pull request merged**